core/src/test/scala/org/apache/pekko/persistence/jdbc/query/JournalDaoStreamMessagesMemoryTest.scala

```diff
@@ -14,138 +14,126 @@
-
-        probe.request(10)
-        probe.within(20.seconds) {
-          probe.expectNextN(10)
-        }
-
-        // sleep and gc to have some kind of stable measurement of current heap usage
-        Thread.sleep(2000)
-        System.gc()
-        Thread.sleep(1000)
-        val usedAfter = memoryMBean.getHeapMemoryUsage.getUsed
-
-        log.info(s"Used heap before ${usedBefore / MB} MB, after ${usedAfter / MB} MB")
-        // actual usage is much less than 10 MB
-        (usedAfter - usedBefore) should be <= (10L * MB)
-
-        probe.cancel()
+      val numberOfInsertBatches = {
+        // calculate the number of batches using a factor to make sure we go a little bit over the limit
+        (maxMem / (payloadSize * eventsPerBatch) * 1.2).round.toInt
       }
+      val totalMessages = numberOfInsertBatches * eventsPerBatch
+      val totalMessagePayload = totalMessages * payloadSize
+      log.info(
+        s"batches: $numberOfInsertBatches (with $eventsPerBatch events), total messages: $totalMessages, total msgs size: $totalMessagePayload")
+
+      // payload can be the same when inserting to avoid unnecessary memory usage
+      val payload = Array.fill(payloadSize)('a'.toByte)
+
+      val lastInsert =
+        Source
+          .fromIterator(() => (1 to numberOfInsertBatches).iterator)
+          .mapAsync(1) { i =>
+            val end = i * eventsPerBatch
+            val start = end - (eventsPerBatch - 1)
+            log.info(s"batch $i - events from $start to $end")
+            val atomicWrites =
+              (start to end).map { j =>
+                AtomicWrite(immutable.Seq(PersistentRepr(payload, j, persistenceId, writerUuid = writerUuid)))
+              }
+            dao.asyncWriteMessages(atomicWrites).map(_ => i)
+          }
+          .runWith(Sink.last)
+
+      // wait until we write all messages
+      // being very generous, 1 second per message
+      lastInsert.futureValue(Timeout(totalMessages.seconds))
+
+      log.info("Events written, starting replay")
+
+      // sleep and gc to have some kind of stable measurement of current heap usage
+      Thread.sleep(1000)
+      System.gc()
+      Thread.sleep(1000)
+      val usedBefore = memoryMBean.getHeapMemoryUsage.getUsed
+
+      val messagesSrc =
+        dao.messagesWithBatch(persistenceId, 0, totalMessages, batchSize = fetchSize, None)
+      val probe =
+        messagesSrc
+          .map {
+            case Success((repr, _)) =>
+              if (repr.sequenceNr % 100 == 0)
+                log.info(s"fetched: ${repr.persistenceId} - ${repr.sequenceNr}/$totalMessages")
+            case Failure(exception) =>
+              log.error("Failure when reading messages.", exception)
+          }
+          .runWith(TestSink.probe)
+
+      probe.request(10)
+      probe.within(20.seconds) {
+        probe.expectNextN(10)
+      }
+
+      // sleep and gc to have some kind of stable measurement of current heap usage
```

Comment:
could this log info be surrounded by a condition that is normally disabled? this logging takes a few seconds and noone is going to look at it

Reply:
ok - I see there is an assertion - this does look like it could be problematic - any tests that have magic sleep times are liable to fail occasionally depending on machine performance